ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/src/Stages/PatternTypeHintsStage.fs

```diff
@@ -0,0 +1,257 @@
+            let symbol = reference.GetFcsSymbol().As<FSharpActivePatternCase>()
+            if isNull symbol then ValueNone
+            else getHintForPattern (asPat.RightPattern.IgnoreInnerParens()) pushToHintMode
+
+        | :? IReferencePat as refPat ->
+            let symbolUse = refPat.GetFcsSymbolUse()
+            if isNull symbolUse then ValueNone else
+            let symbol = symbolUse.Symbol.As<FSharpMemberOrFunctionOrValue>()
+            if isNull symbol then ValueNone else
+
+            let fcsType = symbol.FullType
+            let range = pattern.GetNavigationRange().EndOffsetRange()
+            let displayContext = symbolUse.DisplayContext
+
+            createTypeHintHighlighting fcsType displayContext range pushToHintMode |> ValueSome
+
+        | _ -> ValueNone
+
+    let rec getHighlighting (node: ITreeNode) pushToHintMode =
+        match node with
+        | :? IFSharpPattern as pattern ->
+            getHintForPattern pattern pushToHintMode
+
+        | :? IParameterOwnerMemberDeclaration as decl ->
+            getReturnTypeHint decl pushToHintMode
+
+        | :? IAutoPropertyDeclaration as decl ->
+            let symbolUse = decl.GetFcsSymbolUse()
+            if isNull symbolUse then ValueNone else
+            let symbol = symbolUse.Symbol.As<FSharpMemberOrFunctionOrValue>()
+            if isNull symbol then ValueNone else
+
+            let fcsType = symbol.ReturnParameter.Type
+            let range = decl.NameIdentifier.GetNavigationRange().EndOffsetRange()
+            let displayContext = symbolUse.DisplayContext.WithShortTypeNames(true)
+
+            createTypeHintHighlighting fcsType displayContext range pushToHintMode |> ValueSome
+
+        | _ -> ValueNone
+
+    let adornNodes logKey (topLevelNodes : ITreeNode array) (localNodes : ITreeNode array) =
+        use _swc = logger.StopwatchCookie($"Adorning %s{logKey} nodes", $"topLevelNodesCount=%d{topLevelNodes.Length} localNodesCount=%d{localNodes.Length} sourceFile=%s{daemonProcess.SourceFile.Name}")
+        let highlightingConsumer = FilteringHighlightingConsumer(daemonProcess.SourceFile, fsFile, settingsStore)
+
+        let adornNodes nodes pushToHintMode =
+            for node in nodes do
+                if daemonProcess.InterruptFlag then raise <| OperationCanceledException()
+
+                match getHighlighting node pushToHintMode with
+                | ValueSome highlighting -> highlightingConsumer.AddHighlighting(highlighting)
+                | _ -> ()
+
+        adornNodes topLevelNodes settings.TopLevelMembers
+        adornNodes localNodes settings.LocalBindings
+
+        highlightingConsumer.CollectHighlightings()
+
+    override x.Execute(committer) =
+        let consumer = { TopLevelNodes = List(); LocalNodes = List() }
+        fsFile.Accept(MembersVisitor(settings), consumer)
+
+        let topLevelNodes = consumer.TopLevelNodes |> Array.ofSeq
+        let localNodes = consumer.LocalNodes |> Array.ofSeq
```

Comment:
Could you explain why the arrays are needed, please?

Reply:
Because of `Array.partition` function. Made by analogy with

https://github.com/JetBrains/resharper-fsharp/blob/f6475da23004a3eb8f2487e8a161e11ead450c26/ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/src/Stages/PipeChainTypeHintStage.fs#L150-L164

Perhaps there is a more efficient way, but I didn’t immediately come up with a more elegant solution

Reply:
I'm fine with merging it as is for now, but we should really think about rewriting this and that pieces of code.